lib/puppet_x/templates/invoke_ps_command.erb

```diff
@@ -0,0 +1,284 @@
+$hostSource = @"
+using System;
+using System.Globalization;
+using System.Management.Automation.Host;
+using System.Threading;
+
+public class PuppetPSHost : PSHost
+{
+    private Guid _hostId = Guid.NewGuid();
+    private bool shouldExit;
+    private int exitCode;
+
+    public PuppetPSHost () {}
+
+    public bool ShouldExit
+    {
+        get { return this.shouldExit; }
+    }
+
+    public int ExitCode
+    {
+        get { return this.exitCode; }
+    }
+
+    public void ResetExitStatus()
+    {
+      this.exitCode = 0;
+      this.shouldExit = false;
+    }
+
+    public override Guid InstanceId
+    {
+        get { return _hostId; }
+    }
+
+    public override string Name
+    {
+        get { return "PuppetPSHost"; }
+    }
+
+    public override Version Version
+    {
+        get { return new Version(1, 0); }
+    }
+
+    public override PSHostUserInterface UI
+    {
+        get { return null; }
+    }
+
+    public override CultureInfo CurrentCulture
+    {
+        get { return Thread.CurrentThread.CurrentCulture; }
+    }
+
+    public override CultureInfo CurrentUICulture
+    {
+        get { return Thread.CurrentThread.CurrentUICulture; }
+    }
+
+    public override void EnterNestedPrompt()
+    {
+        throw new NotImplementedException();
+    }
+
+    public override void ExitNestedPrompt()
+    {
+        throw new NotImplementedException();
+    }
+
+    public override void NotifyBeginApplication()
+    {
+        return;
+    }
+
+    public override void NotifyEndApplication()
+    {
+        return;
+    }
+
+    public override void SetShouldExit(int exitCode)
+    {
+      this.shouldExit = true;
+      this.exitCode = exitCode;
+    }
+}
+"@
+
+function New-XmlResult
+{
+  param(
+    [Parameter()]$exitcode,
+    [Parameter()]$output,
+    [Parameter()]$errormessage
+  )
+
+  # we make our own xml because ConvertTo-Xml makes hard to parse xml ruby side
+  # and we need to be sure
+  [System.Xml.XmlDocument]$xDoc = New-Object System.Xml.XmlDocument
```

Comment:
That was the original code, until we ran into problems with non XML escaped text. Then we went the CDATA route, using tags. This didn't escape everything, Glenn suggested the CDATA methods on the XDocument object, so built everything by hand. XDocument outputted utf-16 XML regardless of the docs saying default is utf8 (http://www.timvw.be/2007/01/08/generating-utf-8-with-systemxmlxmlwriter/), thus all the setup and XMLWriter stuff. That still didn't help with all possible characters, but most things. Newlines kept creeping in inside the XML tags, causing XML parsing issues, thus the attempts at indenting to avoid wrapping. Wrapping turned out to be from PSHost not XMLWriter. After abandoning CDATA for Base64, all issues were resolved, but XML left as is when issues with different Ruby and Puppet versions surfaced.